Assets/Scripts/Layers/Adapters/DataTypeAdapters/WFSImportAdapter.cs

```diff
@@ -0,0 +1,292 @@
+                geoJsonOutputFormat = GetGeoJSONOutputFormat(getFeatureOperationNode, namespaceManager);
+                if (string.IsNullOrEmpty(geoJsonOutputFormat))
+                {
+                    Debug.Log("<color=orange>WFS GetFeature operation does not support GeoJSON output format.</color>");
+                    return false;
+                }
+
+                // Get the WFS version
+                wfsVersion = GetWFSVersion(sourceUrl);
+                if(string.IsNullOrEmpty(wfsVersion))
+                    wfsVersion = GetWFSVersion(xmlDocument, namespaceManager);
+            }
+
+            if (getFeatureRequest)
+            {
+                // Get wfs version from url
+                wfsVersion = GetWFSVersion(sourceUrl);
+                geoJsonOutputFormat = "geojson";
+
+                //Check if text is GeoJSON by trying to parse feature collection
+                var featureCollection = JsonConvert.DeserializeObject<FeatureCollection>(dataAsText);
+                if(featureCollection == null)
+                {
+                    Debug.Log("<color=orange>WFS GetFeature request does not return a FeatureCollection.</color>");
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public void Execute(LocalFile localFile)
+        {
+            var sourceUrl = localFile.SourceUrl;
+
+            var getCapabilitiesRequest = sourceUrl.ToLower().Contains("request=getcapabilities");
+            var wfsFolder = new FolderLayer(sourceUrl);
+
+            if(getCapabilitiesRequest)
+            {
+                var featureTypes = GetFeatureTypes(localFile);
+
+                //Create a folder layer 
+                foreach (var featureType in featureTypes)
+                {
+                    Debug.Log("Adding WFS layer for featureType: " + featureType);
+                    AddWFSLayer(featureType, sourceUrl, wfsFolder);
+                }
+                return;
+            }
+
+            var getFeatureRequest = sourceUrl.ToLower().Contains("request=getfeature");
+            if(getFeatureRequest)
+            {
+                // Get the feature type from the url
+                var featureType = string.Empty;
+                if (sourceUrl.ToLower().Contains("typename="))
+                {
+                    //WFS 1.0.0 uses 'typename'
+                    featureType = sourceUrl.ToLower().Split("typename=")[1].Split("&")[0];
+                }
+                else if (sourceUrl.ToLower().Contains("typenames="))
+                {
+                    //WFS 2 uses plural 'typenames'
+                    featureType = sourceUrl.ToLower().Split("typenames=")[1].Split("&")[0];
+                }
+                AddWFSLayer(featureType, sourceUrl, wfsFolder);
+                return;
+            }
+        }
+
+
+        private XmlNamespaceManager ReadNameSpaceManager(XmlDocument xmlDocument)
```

Comment:
I would move some of these checks into a dedicated service instead of the adapter, to keep that lean

Reply:
True; agreed!

Reply:
Nested class (specific to WFS adapter; but cleaner responsibility)